ejenti/server_jobs/backfill_trigger.py

```diff
@@ -0,0 +1,234 @@
+import os
+import json
+import random
+import logging
+from datetime import datetime, timedelta
+
+from lib.common.statusFileCreator import StatusFileCreator
+from lib.helper.generateBackfillTableHelper import GenerateBackfillTableHelper
+from lib.modules.build_information import BuildInformation
+from lib.common.commonUtil import CommonUtil
+from ejenti.pulse_modules.hasalPulsePublisher import HasalPulsePublisher
+
+
+class BackFillTrigger(object):
+
+    PLATFORM = 'win64'
+    RESULT_AMOUNT = 6
+    WAIT_TIMEOUT_DELTA_HOUR = 8
+
+    DEFAULT_CMD_CONFIG_FILE_PATH = os.path.join('configs', 'ejenti', 'cmd_config.json')
+
+    DEFAULT_CMD = 'run-hasal-on-specify-nightly'
+    KEY_OVERWRITE_HASAL_SUITE_CASE_LIST = 'OVERWRITE_HASAL_SUITE_CASE_LIST'
+    KEY_DOWNLOAD_PKG_DIR_URL = 'DOWNLOAD_PKG_DIR_URL'
+    KEY_DOWNLOAD_REVISION = 'DOWNLOAD_REVISION'
+
+    KEY_PERFHERDER_DATA = 'perfherder_data'
+    KEY_VALUE = 'value'
+
+    BROWSER_FIREFOX = 'firefox'
+    BROWSER_CHROME = 'chrome'
+
+    PLATFORM_WIN7 = 'windows8-64'
+    PLATFORM_WIN10 = 'windows10-64'
+
+    PLATFORM_MAPPING = {
+        'windows8-64': 'win7',
+        'windows10-64': 'win10'
+    }
+
+    def __init__(self):
+        self.pulse_username = ''
+        self.pulse_password = ''
+        self.pulse_configs = {}
+        self.casename_set = set()
+
+    def send_task_to_pulse(self, job_name, topic, real_casename, amount, build_info):
+        job_id = StatusFileCreator.create_job_id_folder(job_name)
+        job_id_fp = os.path.join(StatusFileCreator.get_status_folder(), job_id)
+        # Recording Status
+        StatusFileCreator.create_status_file(job_id_fp, StatusFileCreator.STATUS_TAG_PULSE_TRIGGER_BACKFILL, 100)
+
+        # check queue
+        queue_exists = HasalPulsePublisher.check_pulse_queue_exists(username=self.pulse_username,
+                                                                    password=self.pulse_password,
+                                                                    topic=topic)
+
+        if not queue_exists:
+            logging.error('There is not Queue for Topic [{topic}]. Message might be ignored.'.format(topic=topic))
+
```

Comment:
if the queue does not exist, do we need to continue the following steps?

Reply:
It's a trade-off between cpu resource and case running.
Not only publisher can create queue, any one of consumers (agents) also can create queue.
It still has chance to submit tasks after queue be created by consumer.